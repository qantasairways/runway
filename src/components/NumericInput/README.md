### NumericInput

```js
const isBadValue = value => value === 6;

<StatefulManager initial={4}>
  {({ value, updater }) => (
    <NumericInput
      label="Label content"
      id="manual-id-1"
      min={0}
      max={12}
      onChange={updater}
      value={value}
      highlightInvalid={isBadValue(value)}
    />
  )}
</StatefulManager>;
```
