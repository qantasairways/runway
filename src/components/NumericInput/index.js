/* eslint-disable react/jsx-one-expression-per-line */
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import InputNumber from 'rc-input-number';

import PlusIcon from '../../icons/PlusIcon';
import MinusIcon from '../../icons/MinusIcon';

import { colours, mq, highlightInvalidField } from '../../theme/airways';
import { toCx, forAll } from '../../utils/css';

const SELECTORS = {
  RCI: {
    ROOT: {
      CORE: '.rc-input-number',
      FOCUSED: '.rc-input-number-focused',
      DISABLED: '.rc-input-number-disabled'
    },
    CONTROLS: {
      WRAPPER: 'rc-input-number-handler-wrap',
      UP: {
        CORE: '.rc-input-number-handler-up',
        DISABLED: '.rc-input-number-handler-up-disabled',
        ICON: '.runway-numeric-input__increment_icon'
      },
      DOWN: {
        CORE: '.rc-input-number-handler-down',
        DISABLED: '.rc-input-number-handler-down-disabled',
        ICON: '.runway-numeric-input__decrement_icon'
      },
      UP_AND_DOWN: {
        CORE: '.rc-input-number-handler',
        ACTIVE: '.rc-input-number-handler-active'
      }
    },
    INPUT: {
      WRAPPER: '.rc-input-number-input-wrap',
      CORE: '.rc-input-number-input'
    }
  }
};

const iconWidth = '22px';

const getDisabledCtrlStyles = ({ isInvalid }) =>
  isInvalid ? { fill: colours.lighterGrey } : { fill: colours.grey };

const disableNativeNumberInputStyles = {
  [forAll(
    'input::-webkit-outer-spin-button',
    'input::-webkit-inner-spin-button'
  )]: {
    '-webkit-appearance': 'none',
    margin: '0px'
  },
  'input[type="number"]': {
    '-moz-appearance': 'textfield'
  }
};

const getRcInputNumberStyles = ({ isInvalid }) => ({
  ...disableNativeNumberInputStyles,
  [SELECTORS.RCI.ROOT.CORE]: {
    margin: '0',
    padding: '0',
    lineHeight: '26px',
    fontSize: '12px',
    display: 'inline-block',
    verticalAlign: 'middle',
    border: `2px solid ${colours.grey}`,
    borderRadius: '0px',
    position: 'relative',
    width: '100%',
    height: '55px',
    ...(isInvalid && { ...highlightInvalidField })
  },
  [SELECTORS.RCI.ROOT.FOCUSED]: {
    borderColor: '#8de2e0',
    ...(isInvalid && { ...highlightInvalidField })
  },
  [SELECTORS.RCI.CONTROLS.UP_AND_DOWN.CORE]: {
    textAlign: 'center',
    lineHeight: '12px',
    height: '12px',
    overflow: 'hidden',
    display: 'block',
    touchAction: 'none'
  },
  [SELECTORS.RCI.CONTROLS.UP.CORE]: {
    position: 'absolute',
    right: '0px',
    width: `calc(30px + ${iconWidth})`,
    height: '100%'
  },
  [SELECTORS.RCI.CONTROLS.DOWN.CORE]: {
    position: 'absolute',
    left: '0px',
    width: `calc(30px + ${iconWidth})`,
    height: '100%'
  },
  [SELECTORS.RCI.INPUT.CORE]: {
    textAlign: 'center',
    outline: '0',
    '-moz-appearance': 'textfield',
    lineHeight: '26px',
    height: '100%',
    transition: 'all 0.3s ease',
    border: '0',
    padding: '0',
    width: '99%',
    borderRadius: '0px',
    display: 'block',
    fontSize: '22px',
    color: '#323232',
    fontFamily: 'Ciutadella'
  },
  [SELECTORS.RCI.INPUT.WRAPPER]: {
    overflow: 'hidden',
    height: '100%'
  },
  [forAll(
    SELECTORS.RCI.CONTROLS.DOWN.DISABLED,
    SELECTORS.RCI.CONTROLS.UP.DISABLED
  )]: {
    opacity: '0.72'
  },
  [forAll(
    SELECTORS.RCI.CONTROLS.DOWN.DISABLED,
    SELECTORS.RCI.CONTROLS.UP.DISABLED
  )]: {
    [forAll(
      SELECTORS.RCI.CONTROLS.UP.ICON,
      SELECTORS.RCI.CONTROLS.DOWN.ICON
    )]: {
      ...getDisabledCtrlStyles({ isInvalid })
    }
  }
});

const labelStyles = {
  height: '28px',
  lineHeight: '1.75',
  letterSpacing: 'normal',
  fontSize: '16px',
  fontFamily: 'Ciutadella',
  color: colours.darkerGrey
};

const axValidationContainerStyles = {
  position: 'absolute !important',
  display: 'block',
  visibility: 'visible',
  overflow: 'hidden',
  width: '1px',
  height: '1px',
  margin: '-1px',
  border: '0',
  padding: '0',
  clip: 'rect(0 0 0 0)'
};

const axValidationStyles = {
  label: 'runway-numeric-input__error-message',
  marginTop: '15px',
  padding: '13px 15px',
  a: {
    color: colours.darkerGrey,
    textDecoration: 'underline'
  },
  backgroundColor: '#fcebcd',
  fontFamily: 'Ciutadella',
  fontSize: '14px',
  position: 'relative',
  '&:after': {
    content: '""',
    position: 'absolute',
    top: '0',
    left: '50%',
    width: '0',
    height: '0',
    border: '0.6em solid transparent',
    borderBottomColor: '#fcebcd',
    borderTop: '0',
    marginLeft: '-45%',
    marginTop: '-0.531em',
    [mq.medium]: {
      left: '19px',
      marginLeft: '0'
    }
  }
};
class NumericInput extends Component {
  state = {
    value: this.props.value
  };

  componentDidMount = () => {
    const clickables = document.querySelectorAll(
      SELECTORS.RCI.CONTROLS.UP_AND_DOWN.CORE
    );
    clickables.forEach(el => {
      el.addEventListener('click', this.preventDefault);
    });

    if (typeof this.props.setRef === 'function' && this.inputRef) {
      this.props.setRef(this.inputRef);
    }
  };

  overloadedOnChange = value => {
    this.setState({ value });
    this.props.onChange(value);
  };

  componentWillUnmount = () => {
    const clickables = document.querySelectorAll(
      SELECTORS.RCI.CONTROLS.UP_AND_DOWN.CORE
    );
    clickables.forEach(el => {
      el.removeEventListener('click', this.preventDefault);
    });
  };

  preventDefault = e => {
    e.preventDefault();
    e.stopPropagation();
  };

  setInputRef = el => {
    this.inputRef = el;
  };

  render = () => {
    const {
      label,
      id,
      isInvalid,
      isInvalidMessage,
      onChange,
      ...rest
    } = this.props;

    const up = (
      <PlusIcon
        className={toCx(SELECTORS.RCI.CONTROLS.UP.ICON)}
        height="100%"
        width="30px"
      />
    );
    const down = (
      <MinusIcon
        className={toCx(SELECTORS.RCI.CONTROLS.DOWN.ICON)}
        height="100%"
        width="30px"
      />
    );
    return (
      <Fragment>
        {/* eslint-disable-next-line jsx-a11y/label-has-for */}
        <label htmlFor={id} css={labelStyles}>
          {label}
          <div css={getRcInputNumberStyles({ isInvalid })}>
            <InputNumber
              {...rest}
              type="number"
              id={id}
              onChange={this.overloadedOnChange}
              upHandler={up}
              downHandler={down}
              ref={this.setInputRef}
            />
          </div>
        </label>
        <div aria-live="polite" aria-atomic="true">
          <div css={axValidationContainerStyles}>
            Current value is {this.state.value}.
          </div>
          {isInvalid && isInvalidMessage && (
            <div css={axValidationStyles}>{isInvalidMessage}</div>
          )}
        </div>
      </Fragment>
    );
  };
}

NumericInput.propTypes = {
  ...InputNumber.propTypes,
  isInvalid: PropTypes.bool,
  isInvalidMessage: PropTypes.string,
  setRef: PropTypes.func
};

NumericInput.defaultProps = {
  isInvalid: false,
  isInvalidMessage: null,
  setRef: null
};

export default NumericInput;
