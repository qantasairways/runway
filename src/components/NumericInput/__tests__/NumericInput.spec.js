import React from 'react';
import { mount } from 'enzyme';
import NumericInput from '..';

jest.mock('../../../icons/PlusIcon', () => () => <span>Mock Plus Icon</span>);
jest.mock('../../../icons/MinusIcon', () => () => <span>Mock Minus Icon</span>);

describe('NumericInput', () => {
  it('renders correctly with defaults', () => {
    const component = mount(<NumericInput />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props', () => {
    const additionalProps = {
      isInvalid: true,
      isInvalidMessage: 'mock message',
      isTabletAndUp: false
    };
    const component = mount(<NumericInput {...additionalProps} />);
    expect(component).toMatchSnapshot();
  });
});
