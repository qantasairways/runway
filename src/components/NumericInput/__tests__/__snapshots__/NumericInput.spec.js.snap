// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput renders correctly with defaults 1`] = `
.emotion-1 {
  height: 28px;
  line-height: 1.75;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  font-size: 16px;
  font-family: Ciutadella;
  color: #323232;
}

.emotion-0 input::-webkit-outer-spin-button,
.emotion-0 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0px;
}

.emotion-0 input[type="number"] {
  -moz-appearance: textfield;
}

.emotion-0 .rc-input-number {
  margin: 0px;
  padding: 0px;
  line-height: 26px;
  font-size: 12px;
  display: inline-block;
  vertical-align: middle;
  border: 2px solid #dadada;
  border-radius: 0px;
  position: relative;
  width: 100%;
  height: 55px;
}

.emotion-0 .rc-input-number-focused {
  border-color: #8de2e0;
}

.emotion-0 .rc-input-number-handler {
  text-align: center;
  line-height: 12px;
  height: 12px;
  overflow: hidden;
  display: block;
  touch-action: none;
}

.emotion-0 .rc-input-number-handler-up {
  position: absolute;
  right: 0px;
  width: calc(30px + 22px);
  height: 100%;
}

.emotion-0 .rc-input-number-handler-down {
  position: absolute;
  left: 0px;
  width: calc(30px + 22px);
  height: 100%;
}

.emotion-0 .rc-input-number-input {
  text-align: center;
  outline: 0px;
  -moz-appearance: textfield;
  line-height: 26px;
  height: 100%;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  border: 0px;
  padding: 0px;
  width: 99%;
  border-radius: 0px;
  display: block;
  font-size: 22px;
  color: #323232;
  font-family: Ciutadella;
}

.emotion-0 .rc-input-number-input-wrap {
  overflow: hidden;
  height: 100%;
}

.emotion-0 .rc-input-number-handler-down-disabled .runway-numeric-input__increment_icon,
.emotion-0 .rc-input-number-handler-up-disabled .runway-numeric-input__increment_icon,
.emotion-0 .rc-input-number-handler-down-disabled .runway-numeric-input__decrement_icon,
.emotion-0 .rc-input-number-handler-up-disabled .runway-numeric-input__decrement_icon {
  fill: #dadada;
}

.emotion-2 {
  position: absolute !important;
  display: block;
  visibility: visible;
  overflow: hidden;
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0px;
  padding: 0px;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
}

<Fragment>
  <label
    className="emotion-1"
  >
    <div
      className="emotion-0"
    >
      <InputNumber
        autoComplete="off"
        downHandler={
          <MinusIcon
            className="runway-numeric-input__decrement_icon"
            height="100%"
            width="30px"
          />
        }
        focusOnUpDown={true}
        min={-9007199254740991}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        parser={[Function]}
        prefixCls="rc-input-number"
        required={false}
        setRef={null}
        step={1}
        style={Object {}}
        type="number"
        upHandler={
          <PlusIcon
            className="runway-numeric-input__increment_icon"
            height="100%"
            width="30px"
          />
        }
        useTouch={false}
      />
    </div>
  </label>
  <div
    aria-atomic="true"
    aria-live="polite"
  >
    <div
      className="emotion-2"
    >
      Current value is 
      .
    </div>
  </div>
</Fragment>
`;

exports[`NumericInput renders correctly with props 1`] = `
.emotion-1 {
  height: 28px;
  line-height: 1.75;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  font-size: 16px;
  font-family: Ciutadella;
  color: #323232;
}

.emotion-2 {
  position: absolute !important;
  display: block;
  visibility: visible;
  overflow: hidden;
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0px;
  padding: 0px;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
}

.emotion-0 input::-webkit-outer-spin-button,
.emotion-0 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0px;
}

.emotion-0 input[type="number"] {
  -moz-appearance: textfield;
}

.emotion-0 .rc-input-number {
  margin: 0px;
  padding: 0px;
  line-height: 26px;
  font-size: 12px;
  display: inline-block;
  vertical-align: middle;
  border: 2px solid #dadada;
  border-radius: 0px;
  position: relative;
  width: 100%;
  height: 55px;
  border-color: #ed710b;
}

.emotion-0 .rc-input-number-focused {
  border-color: #ed710b;
}

.emotion-0 .rc-input-number-handler {
  text-align: center;
  line-height: 12px;
  height: 12px;
  overflow: hidden;
  display: block;
  touch-action: none;
}

.emotion-0 .rc-input-number-handler-up {
  position: absolute;
  right: 0px;
  width: calc(30px + 22px);
  height: 100%;
}

.emotion-0 .rc-input-number-handler-down {
  position: absolute;
  left: 0px;
  width: calc(30px + 22px);
  height: 100%;
}

.emotion-0 .rc-input-number-input {
  text-align: center;
  outline: 0px;
  -moz-appearance: textfield;
  line-height: 26px;
  height: 100%;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  border: 0px;
  padding: 0px;
  width: 99%;
  border-radius: 0px;
  display: block;
  font-size: 22px;
  color: #323232;
  font-family: Ciutadella;
}

.emotion-0 .rc-input-number-input-wrap {
  overflow: hidden;
  height: 100%;
}

.emotion-0 .rc-input-number-handler-down-disabled .runway-numeric-input__increment_icon,
.emotion-0 .rc-input-number-handler-up-disabled .runway-numeric-input__increment_icon,
.emotion-0 .rc-input-number-handler-down-disabled .runway-numeric-input__decrement_icon,
.emotion-0 .rc-input-number-handler-up-disabled .runway-numeric-input__decrement_icon {
  fill: #eeeeee;
}

.emotion-3 {
  margin-top: 15px;
  padding: 13px 15px;
  background-color: #fcebcd;
  font-family: Ciutadella;
  font-size: 14px;
  position: relative;
}

.emotion-3 a {
  color: #323232;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-3:after {
  content: "";
  position: absolute;
  top: 0px;
  left: 50%;
  width: 0px;
  height: 0px;
  border: 0.6em solid transparent;
  border-bottom-color: #fcebcd;
  border-top: 0px;
  margin-left: -45%;
  margin-top: -0.531em;
}

@media only screen and (min-width:768px) {
  .emotion-3:after {
    left: 19px;
    margin-left: 0px;
  }
}

<Fragment>
  <label
    className="emotion-1"
  >
    <div
      className="emotion-0"
    >
      <InputNumber
        autoComplete="off"
        downHandler={
          <MinusIcon
            className="runway-numeric-input__decrement_icon"
            height="100%"
            width="30px"
          />
        }
        focusOnUpDown={true}
        min={-9007199254740991}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        parser={[Function]}
        prefixCls="rc-input-number"
        required={false}
        setRef={null}
        step={1}
        style={Object {}}
        type="number"
        upHandler={
          <PlusIcon
            className="runway-numeric-input__increment_icon"
            height="100%"
            width="30px"
          />
        }
        useTouch={false}
      />
    </div>
  </label>
  <div
    aria-atomic="true"
    aria-live="polite"
  >
    <div
      className="emotion-2"
    >
      Current value is 
      .
    </div>
    <div
      className="emotion-3"
    >
      mock message
    </div>
  </div>
</Fragment>
`;
